core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/ruin/RuinPhase.java

```diff
@@ -0,0 +1,129 @@
+package org.optaplanner.core.impl.ruin;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.impl.phase.AbstractPhase;
+import org.optaplanner.core.impl.phase.custom.CustomPhase;
+import org.optaplanner.core.impl.phase.custom.scope.CustomPhaseScope;
+import org.optaplanner.core.impl.phase.custom.scope.CustomStepScope;
+import org.optaplanner.core.impl.phase.scope.AbstractPhaseScope;
+import org.optaplanner.core.impl.phase.scope.AbstractStepScope;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+import org.optaplanner.core.impl.solver.termination.Termination;
+
+/**
+ * Default implementation of {@link RuinPhase}.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ */
+public class RuinPhase<Solution_> extends AbstractPhase<Solution_> implements CustomPhase<Solution_> {
```

Comment:
I don't think the entire phase is necessary. Here's what I propose:

- The ruin config creates a custom phase.
- The custom phase gets the ruin command.

As a result, the logic of running a phase stays in custom phase, and ruination is only a command for it. That way, code duplication is reduced to a minimum.